### PR TITLE
Make config throw errors and take defaults.

### PR DESCRIPTION
**Pull request by ericnormand**
`config` is now a 2-arity function. It takes the key path and an
optional default value.

If no default is given and the key path does not exist in the config
file, an exception is thrown. I tried to make the exception informative
as to what was expected to exist and where. It contains the key path,
the current config file's value (after reading), and the filename of the
config file.

If a default is given and the key path does not exist, the default is
returned.

We discussed this at length on Slack. The main problem was tracking down
exceptions (often due to `nil`s) that were caused by missing or
misspelled config items. It was pretty unanimous that at least a warning
should be issued if a config value was looked for but it didn't exist.

The debate had two sides: one claiming that it should throw an exception
if the value does not exist and one claiming that this was drastic and
the code reading from the config should use idiomatic `(or (config :x)
default)` or `(if-let [val (config :x)]...` constructs.

This is an implementation of the former side. There is still a single
function with two arities. The choice is obvious: if you have a default,
put it, otherwise, don't. If you want to avoid the exception, put a
default (could be `nil`). Either way everything is explicit.

Finally, the tests were modified to use the new key path style (using
vectors). New tests were added to test the exception and default
behavior.

### CODE REVIEW

**Comment by ericnormand** [comment]
Assigning to @danielglauser because I know @tie-rack is going to read it anyway.

File: src/turbovote/resource_config.clj [hunk 1]
```diff
@@ -14,5 +14,37 @@
...
+                :config-filename (-> config-file-name
+                                     io/resource
+                                     io/file
+                                     str)})))
+     val))
+  ([keys default]
+   (let [val (get-in (read-config config-file-name) keys ::not-found)]
+     (if (= ::not-found val)
+       default
+       val))))
```
Comment (tie-rack) [inline src/turbovote/resource_config.clj]: Couldn't the 2-arity version just be `(get-in (read-config config-file-name) keys default)` ?

File: src/turbovote/resource_config.clj [hunk 2]
```diff
@@ -14,5 +14,37 @@
...
+       (throw (ex-info
+               (str "Hey! I went looking for these keys ("
+                    (pr-str keys)
+                    ") in the config "
+                    "and I couldn't find them.")
+               {:keys keys
+                :config config
+                :config-filename (-> config-file-name
+                                     io/resource
+                                     io/file
```
Comment (tie-rack) [inline src/turbovote/resource_config.clj]: The resource is not necessarily going to be able to be coerced into a File (if it's in a jar or war or however a vfs happens), causing `io/file` to throw before we get to throw this.

File: src/turbovote/resource_config.clj (hunk 1 above)
Comment (ericnormand) [inline src/turbovote/resource_config.clj]: Yes. Good catch.

File: src/turbovote/resource_config.clj (hunk 2 above)
Comment (ericnormand) [inline src/turbovote/resource_config.clj]: You're right.

Comment (danielglauser) [inline src/turbovote/resource_config.clj]: If you just returned the resource I'm sure that would be enough. I believe it's string form is just a URI.

File: test/turbovote/resource_config_test.clj [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+(deftest config-default-arity
+  (testing "can add default to config"
+    (is (= (config [:startup-message] "Hey, dude!") "Hello, world!"))
+    (is (= ::default (config [:some :random :path :that :does :not :exist]
+                             ::default)))))
+
+(deftest missing-config-test
+  (is (thrown-with-msg?
+       java.io.FileNotFoundException #"Config file __missing__\.edn not found in resource paths"
```
Comment (danielglauser) [inline test/turbovote/resource_config_test.clj]: Just a reminder that once you remove `io/file` I don't think we'll still get a `java.io.FileNotFoundException`.

**Comment by danielglauser** [comment]
Awesome @ericnormand! Thanks for moving on this so quickly.
LGTM +1

**Comment by tie-rack** [comment]
Great!

**Pull request merged**